### PR TITLE
feat(frontend): popup on profile when user leaves bb-473

### PR DESCRIPTION
**Pull request by MaxKorop**
### Summary

- added a popup when the user leaves the profile page without saving changes with the name
- added a popup when the user leaves the profile page without saving changes with the password

### Screenshots
image

### CODE REVIEW

**Comment by MaxKorop** [comment]
Found some bugs, I'll fix they now

File: apps/frontend/src/pages/profile/profile.tsx [hunk 1]
```diff
@@ -73,6 +74,24 @@ const Profile: React.FC = () => {
 
 	const isLoading = dataStatus === DataStatus.PENDING;
 
+	const [isDirty, setIsDirty] = useState<boolean>(false);
```
Comment (fshabanov) [inline apps/frontend/src/pages/profile/profile.tsx:77]: Not sure to understand why we need this state. `react-hook-form` should handle `isDirty`

File: apps/frontend/src/pages/profile/profile.tsx [hunk 2]
```diff
@@ -73,6 +74,24 @@ const Profile: React.FC = () => {
...
+		if (blocker.state === "blocked") {
+			blocker.reset();
+		}
+	}, [blocker]);
+
+	const handleConfirmPopupClick = useCallback((): void => {
+		if (blocker.state === "blocked") {
+			blocker.proceed();
+		}
+	}, [blocker]);
```
Comment (fshabanov) [inline apps/frontend/src/pages/profile/profile.tsx:93]: I feel like this will be duplicated lots of times, let's create a separate hook for that logic

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/pages/profile/profile.tsx (hunk 1 above)
Comment (MaxKorop) [reply apps/frontend/src/pages/profile/profile.tsx:77]: We need this state to handle two forms in separate components, but `react-router-dom` allows only one blocker per route. For example, after the user changes something in the form of the `UpdateUserForm` component, the blocker and router don't work and the user gets this warning in the console
image

File: apps/frontend/src/libs/hooks/hooks.ts [hunk 3]
```diff
@@ -3,6 +3,10 @@ export { useAppForm } from "./use-app-form/use-app-form.hook.js";
 export { useAppSelector } from "./use-app-selector/use-app-selector.hook.js";
 export { useClickOutside } from "./use-click-outside/use-click-outside.hook.js";
 export { useQuery } from "./use-query/use-query-hook.js";
+export {
+	BLOCKED_BLOCKER_STATE,
```
Comment (fshabanov) [inline apps/frontend/src/libs/hooks/hooks.ts:7]: You shouldn't export a constant from hooks file

File: apps/frontend/src/libs/hooks/use-unsaved-changes-blocker/use-unsaved-changes-blocker.hook.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	reset,
+}: Properties): ReturnType => {
+	const blocker = useBlocker(({ currentLocation, nextLocation }) => {
+		return (
+			hasUncavedChanges && currentLocation.pathname !== nextLocation.pathname
+		);
+	});
+
+	const handlePopupCancel = useCallback((): void => {
+		if (blocker.state === "blocked") {
```
Comment (fshabanov) [inline apps/frontend/src/libs/hooks/use-unsaved-changes-blocker/use-unsaved-changes-blocker.hook.ts:25]: You have "blocked" in a constant

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/hooks/use-unsaved-changes-blocker/use-unsaved-changes-blocker.hook.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import { BLOCKED_BLOCKER_STATE } from "~/libs/constants/blocked-blocker-state.constant.js";
```
Comment (fshabanov) [inline apps/frontend/src/libs/hooks/use-unsaved-changes-blocker/use-unsaved-changes-blocker.hook.ts:1]:
```suggestion
import { BLOCKED_BLOCKER_STATE } from "~/libs/constants/constants.js";
```

File: apps/frontend/src/libs/hooks/use-unsaved-changes-blocker/use-unsaved-changes-blocker.hook.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+	const handlePopupCancel = useCallback((): void => {
+		if (blocker.state === BLOCKED_BLOCKER_STATE) {
+			blocker.reset();
+		}
+	}, [blocker]);
+
+	const handlePopupConfirm = useCallback((): void => {
+		if (reset) {
+			reset();
+		}
```
Comment (fshabanov) [inline apps/frontend/src/libs/hooks/use-unsaved-changes-blocker/use-unsaved-changes-blocker.hook.ts:34]:
```suggestion
			reset?.();
```

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/hooks/use-unsaved-changes-blocker/use-unsaved-changes-blocker.hook.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+	useEffect(() => {
+		window.addEventListener("beforeunload", handleBeforeUnload);
+
+		return (): void => {
+			window.removeEventListener("beforeunload", handleBeforeUnload);
+		};
+	}, [handleBeforeUnload]);
+
+	return {
+		blockerState: blocker.state,
```
Comment (fshabanov) [inline apps/frontend/src/libs/hooks/use-unsaved-changes-blocker/use-unsaved-changes-blocker.hook.ts:58]: Let's return `isBlocked` instead of state

**Review by fshabanov — APPROVED** [review]

File: apps/frontend/src/libs/hooks/use-unsaved-changes-blocker/use-unsaved-changes-blocker.hook.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+		reset?.();
+
+		if (blocker.state === BLOCKED_BLOCKER_STATE) {
+			blocker.proceed();
+		}
+	}, [blocker, reset]);
+
+	const handleBeforeUnload = useCallback(
+		(event: BeforeUnloadEvent): void => {
+			if (hasUncavedChanges) {
```
Comment (fshabanov) [inline apps/frontend/src/libs/hooks/use-unsaved-changes-blocker/use-unsaved-changes-blocker.hook.ts:41]:
```suggestion
			if (hasUnsavedChanges) {
```

**Review by fshabanov — APPROVED** [review]

**Review by fshabanov — APPROVED** [review]

**Pull request merged**